### PR TITLE
Improve vmpooler scheduling logic

### PR DESCRIPTION
**Pull request by mattkirby**
This change implements improved scheduling of pooled VMs by updating the logic used to find the least used host in a target cluster. Without this change host selection is based on VM count, which does not give an indication of VM size or current host utilization.

Additionally, this change implements a queue for migrating VMs at request time. The least used compatible host is selected for migration, which occurs transparently to the consumer, and migration is performed if required.

Without these changes vmpooler may cause uneven load distribution across a cluster of resources since the consumption of VMs is separated from scheduling.

### CODE REVIEW

**Comment by mattkirby** [comment]
This change is ready for review, but is blocked from moving forward until I present the changes next week. Please feel free to offer any review in the interim.

File: spec/vmpooler/pool_manager_spec.rb [hunk 1]
```diff
@@ -228,6 +228,7 @@
       allow(redis).to receive(:smembers).with('vmpooler__running__pool1').and_return([])
       allow(redis).to receive(:smembers).with('vmpooler__completed__pool1').and_return([])
       allow(redis).to receive(:smembers).with('vmpooler__discovered__pool1').and_return([])
+      allow(redis).to receive(:smembers).with('vmpooler__migrating__pool1').and_return([])
```
Comment (rick) [inline spec/vmpooler/pool_manager_spec.rb]: The pool manager has the only remaining hive of "mockist" specs in the pooler (I converted the remaining specs over to non-mockist specs).  The "mockist" specs don't really exercise much of anything, but I stopped short here because rewriting these specs is a pain.

It would not be a blocker for this PR to not put in some real specs for the new functionality, but I would definitely give big :sparkles: / :tada: to the effort.

If you end up wanting to go that route, there are some helpers in https://github.com/puppetlabs/vmpooler/blob/master/spec/helpers.rb that are designed to set up actual vm state in tests, which allows us to then call our methods to manipulate and then assert that things look right in redis.

For example, here's a fairly stateful non-mockist spec: https://github.com/puppetlabs/vmpooler/blob/master/spec/vmpooler/api/v1/vm_template_spec.rb#L173-L184

Anyway, don't worry about it if it's a pain (it probably is).  It's on our (my?) wishlist at some point to have our suite be more real, and to evolve the pool manager so that it has a reasonable domain model reflected in the code and those test helpers (or something similar) migrate down into the code itself.

File: lib/vmpooler/pool_manager.rb [hunk 2]
```diff
@@ -455,6 +455,31 @@ def _check_snapshot_queue
...
 
+    def migrate_vm(vm, pool)
+      Thread.new do
+        _migrate_vm(vm, pool)
+      end
+    end
+
+    def _migrate_vm(vm, pool)
+      $redis.srem('vmpooler__migrating__' + pool, vm)
+      vm_object = $vsphere[pool].find_vm(vm) || $vsphere[pool].find_vm_heavy(vm)
```
Comment (rick) [inline lib/vmpooler/pool_manager.rb]: Again (see below), not a priority to fix here, but if it's easy, this incantation could become a method with a name (e.g., `find_vsphere_pool_vm(pool, vm)`).

File: lib/vmpooler/pool_manager.rb [hunk 3]
```diff
@@ -555,6 +581,17 @@ def _check_pool(pool)
         end
       end
 
+      # MIGRATIONS
+      $redis.smembers('vmpooler__migrating__' + pool['name']).each do |vm|
+        if inventory[vm]
+          begin
+            migrate_vm(vm, pool['name'])
+          rescue
```
Comment (rick) [inline lib/vmpooler/pool_manager.rb]: It might be handy for logging purposes here to make this:

``` ruby
begin
  migrate_vm(vm, pool['name'])
rescue => err
  $logger.log('s', '[x] [' + pool['name'] + "] '" + vm + "' failed to migrate: #{err}")
end
```

The log line could also be written as:

``` ruby
$logger.log('s', "[x] [#{pool['name']}] '#{vm}' failed to migrate: #{err}")
```

File: lib/vmpooler/vsphere_helper.rb [hunk 4]
```diff
@@ -196,31 +196,62 @@ def find_folder(foldername)
...
+          memory_utilization = (memory_usage.to_f / memory_size.to_f) * 100
+          if not cpu_utilization > limit and not memory_utilization > limit
+            utilization = cpu_utilization + memory_utilization
+            return [ utilization, host ]
+          end
+        end
+      end
+      return nil
+    end
+
```
Comment (rick) [inline lib/vmpooler/vsphere_helper.rb]: A potential refactoring of this method (untested):

``` ruby
  def get_host_utilization(host, model=nil, limit=90)
    return nil unless model
    return nil unless host_has_cpu_model? host, model
    return nil if host.runtime.inMaintenanceMode
    return nil unless host.overallStatus == 'green'

    cpu_utilization = cpu_utilization_for host
    memory_utilization = memory_utilization_for host

    return nil if cpu_utilization > limit
    return nil if memory_utilization > limit

    [ cpu_utilization + memory_utilization, host ]
  end

  def host_has_cpu_model?(host, model)
    host.hardware.cpuPkg[0].description.split().include? model
  end

  def cpu_utilization_for(host)
    cpu_usage = host.summary.quickStats.overallCpuUsage
    cpu_size = host.summary.hardware.cpuMhz * host.summary.hardware.numCpuCores
    (cpu_usage.to_f / cpu_size.to_f) * 100
  end

  def memory_utilization_for(host)
    memory_usage = host.summary.quickStats.overallMemoryUsage
    memory_size = host.summary.hardware.memorySize / 1024 / 1024
    (memory_usage.to_f / memory_size.to_f) * 100
  end
```

File: lib/vmpooler/vsphere_helper.rb [hunk 5]
```diff
@@ -196,31 +196,62 @@ def find_folder(foldername)
...
+          host_usage = get_host_utilization(host)
+          target_hosts << host_usage if host_usage
         end
       end
+      return target_hosts.sort[0][1]
+    end
 
-      hosts[hosts_sort.sort_by { |_k, v| v }[0][0]]
+    def find_least_used_compatible_host(vm)
+      begin
```
Comment (rick) [inline lib/vmpooler/vsphere_helper.rb]: This block is a bit cryptic -- I think we're basically calling a thing for the side effect of throwing an error if uninitialized and then initializing in that case.  Probably could extract it to a method and give it an intention-revealing name?

File: lib/vmpooler/vsphere_helper.rb [hunk 6]
```diff
@@ -196,31 +196,62 @@ def find_folder(foldername)
...
+        @connection.serviceInstance.CurrentTime
+      rescue
+        initialize
+      end
+
+      source_host = vm.summary.runtime.host
+      model = source_host.hardware.cpuPkg[0].description.split()[4]
```
Comment (rick) [inline lib/vmpooler/vsphere_helper.rb]: This bit: `host.hardware.cpuPkg[0].description.split()` turns up in multiple places. What is this thing? Can we use the name for the meaning of it, wrap it in a method, and call it?  The `[4]` bit is a bit cryptic. If we extract to a method we could document there what the structure is. Maybe even make a method with a clear name that does the `[4]` part.

File: lib/vmpooler/vsphere_helper.rb [hunk 7]
```diff
@@ -196,31 +196,62 @@ def find_folder(foldername)
...
+      cluster = source_host.parent
+      target_hosts = []
+      cluster.host.each do |host|
+        host_usage = get_host_utilization(host, model)
+        target_hosts << host_usage if host_usage
+      end
+      return target_hosts.sort[0][1]
```
Comment (rick) [inline lib/vmpooler/vsphere_helper.rb]: Nitpick: we don't need `return` at the end of a method in ruby, it's enough just to say `target_hosts.sort[0][1]`.

Also, I notice that we're doing similar things here as we are higher up in the diff.  Makes me think there's a more fundamental operation happening that we could extract to a method and share?

**Comment by rick** [comment]
Mostly style / refactoring / readability / nitpick comments.  Nothing that I mentioned would block this PR.  Am totally game to pair on this if you want to at some point.

File: spec/vmpooler/pool_manager_spec.rb (hunk 1 above)
Comment (mattkirby) [inline spec/vmpooler/pool_manager_spec.rb]: I do want to setup some real tests for the redis usage as I think it will uncover some bugs. I would like to move this change forward separately from that, but I am definitely happy to put in some time to do this since I think it will help identify some of the problem areas of the current redis usage and increase confidence in future changes. The examples are very helpful for me since I have not written much in the way of rub

File: lib/vmpooler/pool_manager.rb (hunk 2 above)
Comment (mattkirby) [inline lib/vmpooler/pool_manager.rb]: I updated this as suggested.

File: lib/vmpooler/pool_manager.rb [hunk 8]
```diff
@@ -455,6 +455,31 @@ def _check_snapshot_queue
...
+    def _migrate_vm(vm, pool)
+      $redis.srem('vmpooler__migrating__' + pool, vm)
+      vm_object = $vsphere[pool].find_vm(vm) || $vsphere[pool].find_vm_heavy(vm)
+      host = $vsphere[pool].find_least_used_compatible_host(vm_object)
+      parent_host = vm_object.summary.runtime.host
+      parent_host_name = parent_host.name
+      if host == parent_host
+        $logger.log('s', '[ ] [' + pool + "] No migration required for '" + vm + "'")
+      else
+        start = Time.now
```
Comment (mattkirby) [inline lib/vmpooler/pool_manager.rb]: This has been added. It looks like it follows the currently used patterns, but please let me know if I've missed something or if it needs further adjustment.

File: lib/vmpooler/pool_manager.rb (hunk 3 above)
Comment (mattkirby) [inline lib/vmpooler/pool_manager.rb]: I have added the context of the failure to the log message and cleaned up the logger string.

File: lib/vmpooler/vsphere_helper.rb (hunk 4 above)
Comment (mattkirby) [inline lib/vmpooler/vsphere_helper.rb]: I ran through and put together something that looks nearly like this and still does the same thing. I think it captures what you were suggesting and helps clean it up. Thank you.

File: lib/vmpooler/vsphere_helper.rb (hunk 5 above)
Comment (mattkirby) [inline lib/vmpooler/vsphere_helper.rb]: Yes. It's a existing pattern that I duplicated here that tries to see if a connection is alive and available for reuse. I moved it to a method and replaced usage of this pattern with a call of the method.

File: lib/vmpooler/vsphere_helper.rb (hunk 6 above)
Comment (mattkirby) [inline lib/vmpooler/vsphere_helper.rb]: I've made an attempt to clarify this with methods. I think it should be more clear now, but am happy to increase clarity if it still requires improvement.

File: lib/vmpooler/pool_manager.rb [hunk 9]
```diff
@@ -455,6 +455,31 @@ def _check_snapshot_queue
       end
     end
 
+    def migrate_vm(vm, pool)
+      Thread.new do
```
Comment (mattkirby) [inline lib/vmpooler/pool_manager.rb]: Yes, a new thread for each VM that needs to be migrated. The limitation ends up being placed by the number of vsphere connections, and there are shared vsphere connections per pool that end up being used (I'm not sure how many). There's a concept of limiting concurrent tasks in vmpooler, but I haven't gotten into the nuts and bolts of how that works yet. I suspect there is some room for optimization. It looks like the model of vmpooler is to always create more threads, so this is consistent with how the application is handling clone operations, and other actions of that nature. The number of operations able to be performed at any one time will always be limited by the number of vsphere connections.

Since the question of which host is the best target for my VM right now is asked for every VM, and the answer is not shared between migration jobs, then it should be able to accurately track the utilization of hosts and deploy new jobs to lesser used hosts minimizing the possibility of hot spots. I would expect a more even distribution of CPU ready time across the cluster to be representative of this.

File: lib/vmpooler/pool_manager.rb [hunk 10]
```diff
@@ -555,6 +581,17 @@ def _check_pool(pool)
...
+      $redis.smembers('vmpooler__migrating__' + pool['name']).each do |vm|
+        if inventory[vm]
+          begin
+            migrate_vm(vm, pool['name'])
+          rescue
+            $logger.log('s', '[x] [' + pool['name'] + "] '" + vm + "' failed to migrate")
+          end
+        end
+      end
+
```
Comment (mattkirby) [inline lib/vmpooler/pool_manager.rb]: What hypervisor data are you hoping to see? The only thing that stands out to me is if you wanted to know the host your VM is running on we could report that when logging that no migration is required. This will already be reported when a migration is performed.

File: lib/vmpooler/vsphere_helper.rb (hunk 7 above)
Comment (mattkirby) [inline lib/vmpooler/vsphere_helper.rb]: Thanks. I wasn't sure what the best practice is and have adjusted this.

File: lib/vmpooler/pool_manager.rb (hunk 10 above)
Comment (mattkirby) [inline lib/vmpooler/pool_manager.rb]: @waynr, I have added this change.

File: lib/vmpooler/vsphere_helper.rb [hunk 11]
```diff
@@ -196,31 +196,62 @@ def find_folder(foldername)
       base
     end
 
+    def get_host_utilization(host, model=nil, limit=90)
```
Comment (mattkirby) [inline lib/vmpooler/vsphere_helper.rb]: I added some comments to try and provide some context for what these means. Let me know if you think further comments are necessary, or would be helpful.

**Comment by mattkirby** [comment]
I have added an additional commit that makes VM migration at checkout optional, defaulting to disabled. When migration_limit is in the config section of vmpooler.yaml, and the value is greater than 0, migration at checkout will happen. In either case this change adds logging to vmpooler.yaml of a VM parent host, whether involving a migration or not.

**Comment by shrug** [comment]
I'm 👍  on getting this in and then tuning. I have not done a deep review, but if @rick has looked it over that's good enough for me

File: vmpooler.yaml.example [hunk 12]
```diff
@@ -225,6 +225,13 @@
 #     If set, prefixes all created VMs with this string.  This should include
 #     a separator.
 #     (optional; default: '')
+#
+#   - migration_limit
+#     When set to any value greateer than 0 enable VM migration at checkout.
```
Comment (rick) [inline vmpooler.yaml.example]: Typo:  "greateer"

**Comment by rick** [comment]
One typo in a comment, otherwise:  LGTM. ✨ :rocket: :guitar: :metal:

**Comment by waynr** [comment]
This looks pretty good to me. @mattkirby is there a particular dashboard we should watch once this change rolls out to production?

File: vmpooler.yaml.example (hunk 12 above)
Comment (mattkirby) [inline vmpooler.yaml.example]: Thanks. Fixed it.

**Comment by rick** [comment]
:+1: from me to move this forward to testing on vmpooler-dev. Could also coordinate with @puppetlabs/cinext about trying it on the cinext vmpooler. If it's looking good there, then prod vmpooler (branch deploy during time of low activity) and eventually merge.

File: lib/vmpooler/vsphere_helper.rb [hunk 13]
```diff
@@ -196,39 +168,98 @@ def find_folder(foldername)
...
-          if
-            (host.overallStatus == 'green') &&
-            (!host.runtime.inMaintenanceMode)
-
-            hosts[host.name] = host
-            hosts_sort[host.name] = host.vm.length
-          end
-        end
+      datacenter.hostFolder.children.each do |cluster_object|
+        return cluster_object if cluster_object.name == cluster
```
Comment (joshcooper) [inline lib/vmpooler/vsphere_helper.rb]: could shorten this to:

```ruby
datacenter.hostFolder.childen.find { |cluster_object| cluster_object.name == cluster }
```

Comment (mattkirby) [inline lib/vmpooler/vsphere_helper.rb]: Thanks. I have shortened this as suggested and updated the PR.

File: lib/vmpooler/vsphere_helper.rb [hunk 14]
```diff
@@ -196,39 +168,96 @@ def find_folder(foldername)
       base
     end
 
-    def find_least_used_host(cluster)
+    def get_host_utilization(host, model=nil, limit=90)
+      # Returns an array containing cumulative CPU and memory utilization of a host, and its object reference
+      # Params:
+      # +model+:: CPU arch version to match on
+      # +limit+:: Hard limit for CPU or memory utilization beyond which a host is excluded for deployments
```
Comment (nwolfe) [inline lib/vmpooler/vsphere_helper.rb]: This should docstring should go above the method signature in order for it to actually be a docstring.  Down here it's just a comment at the top of the body.

File: lib/vmpooler/vsphere_helper.rb [hunk 15]
```diff
@@ -196,39 +168,96 @@ def find_folder(foldername)
...
+      (cpu_usage.to_f / cpu_size.to_f) * 100
+    end
+
+    def memory_utilization_for(host)
+      memory_usage = host.summary.quickStats.overallMemoryUsage
+      memory_size = host.summary.hardware.memorySize / 1024 / 1024
+      (memory_usage.to_f / memory_size.to_f) * 100
+    end
+
+    def vsphere_connection_alive?(connection)
```
Comment (nwolfe) [inline lib/vmpooler/vsphere_helper.rb]: What does this method do?

At first I thought it returned a boolean value since it had the `?` in the name.

But then I saw usages didn't use it like that, so I thought it was a validation method that threw an exception.
But that doesn't appear to be the case either.

File: lib/vmpooler/vsphere_helper.rb (hunk 14 above)
Comment (mattkirby) [inline lib/vmpooler/vsphere_helper.rb]: I have updated it. Hopefully this should look better.

File: lib/vmpooler/vsphere_helper.rb (hunk 15 above)
Comment (nwolfe) [inline lib/vmpooler/vsphere_helper.rb]: > It verifies a connection is alive reconnects when it is not.

Ahh, I couldn't even tell that when looking at the definition itself (I had no idea what that `initialize` was going to do).

I think it would be good to reflect the behavior in the name, so something like `ensure_connected(connection)` or `reconnect_if_dead(connection)` or something like that.

**Comment by nwolfe** [comment]
Is the scheduling logic sophisticated enough to warrant an english description somewhere?

Not trying to imply it is, just wondering.

It can be hard to look at the code and realize the overall behavior.

**Comment by mattkirby** [comment]
Perhaps. It's hard for me to have a good sense of that from my perspective. I am happy to add some documentation if it helps make the intent more clear. It's based on calculated CPU utilization, memory utilization, then summing the two and sorting the list of compatible hosts by their usage to identify the least used.

File: lib/vmpooler/vsphere_helper.rb (hunk 15 above)
Comment (mattkirby) [inline lib/vmpooler/vsphere_helper.rb]: I updated the name of the method. Does the updated version seem more clear? If so, I'll squash the changes.

**Comment by waynr** [comment]
@puppetlabs/cinext so I was just thinking about this and wondered what everyone thinks about the possibility of having vmpooler refuse to check out a VM if it can't find a host with CPU/Memory availability below a certain threshold?

**Comment by rick** [comment]
> @puppetlabs/cinext so I was just thinking about this and wondered what everyone thinks about the possibility of having vmpooler refuse to check out a VM if it can't find a host with CPU/Memory availability below a certain threshold?

If we made a change in that direction I'd prefer that it not make a vm ready rather than refusing the checkout. So the pool would be "empty" in that case.  I think it might be harder to make downstream scheduling decisions (thinking ABS here) if it looks like we have capacity but we have a higher rate of actual checkouts on that "available" capacity failing.

**Comment by waynr** [comment]
Yeah my question came from ignorance (my own fault) of what's currently available in ABS.

As for refusing to make a VM ready rather than refusing a checkout, that would not prevent overutilization of the infrastructure because in that situation there is still a possibility for VMs to be ready well before overutilization begins and to cause catastrophic overutilization by being checked out willy nilly as they are currently. Anyway, setting the limit in ABS definitely suits the need that led me to ask my previous question.

**Comment by mattkirby** [comment]
I've pushed a commit here that adds tests for the migration at checkout functionality. I would like to get some feedback about how the tests are written to ensure I'm testing the right pieces, and that the way the tests are structured are reasonable. I chose to separate the tests to a new file mostly just to help me isolate what I'm testing, but can merge them into a more appropriate place if it seems reasonable. I'm still working on testing the rest, but wanted to submit what I have so far here to ensure I'm on the right path.

File: spec/vmpooler/pool_manager_spec.rb (hunk 1 above)
Comment (mattkirby) [inline spec/vmpooler/pool_manager_spec.rb]: I started adding some specs here and making an effort to test things a little more thoroughly. I have not written rspec tests before, so I would like to get some feedback about how I'm approaching this to identify if I'm on the right path, or if I need to adjust the approach.

File: lib/vmpooler/pool_manager.rb [hunk 16]
```diff
@@ -448,39 +460,111 @@ def _check_snapshot_queue
...
       end
     end
 
+    def find_vsphere_pool_vm(pool, vm, vsphere)
+      vsphere.find_vm(vm) || vsphere.find_vm_heavy(vm)[vm]
+    end
+
+    def migration_limit(migration_limit)
+      # Returns migration_limit setting when enabled
+      return false if migration_limit == 0 or not migration_limit
```
Comment (rick) [inline lib/vmpooler/pool_manager.rb]: In ruby the operator precedence of `||` vs. `or`, and `!` vs. `not` can be surprising.  I *think* this is ok here, but would prefer we use `||` and `!` instead of `or` and `not`.

File: lib/vmpooler/pool_manager.rb [hunk 17]
```diff
@@ -448,39 +460,111 @@ def _check_snapshot_queue
...
+    end
+
+    def _migrate_vm(vm, pool, vsphere)
+      begin
+        $redis.srem('vmpooler__migrating__' + pool, vm)
+        vm_object = find_vsphere_pool_vm(pool, vm, vsphere)
+        parent_host, parent_host_name = get_vm_host_info(vm_object)
+        migration_limit = migration_limit $config[:config]['migration_limit']
+
+        if not migration_limit
```
Comment (rick) [inline lib/vmpooler/pool_manager.rb]: Ditto here for `!` preferred over `not`.

File: lib/vmpooler/pool_manager.rb [hunk 18]
```diff
@@ -448,39 +460,111 @@ def _check_snapshot_queue
...
+          $logger.log('s', "[ ] [#{pool}] '#{vm}' is running on #{parent_host_name}")
+        else
+          migration_count = $redis.smembers('vmpooler__migration').size
```
Comment (rick) [inline lib/vmpooler/pool_manager.rb]: Wondering if `scard` would work here instead of the `smembers().size`.  Also, we might be able to fetch this at the top instead of inline, but I may not understand the intricacies here.

File: lib/vmpooler/pool_manager.rb [hunk 19]
```diff
@@ -448,39 +460,111 @@ def _check_snapshot_queue
...
+          if migration_count >= migration_limit
+            $logger.log('s', "[ ] [#{pool}] '#{vm}' is running on #{parent_host_name}. No migration will be evaluated since the migration_limit has been reached")
+          else
```
Comment (rick) [inline lib/vmpooler/pool_manager.rb]: Purely a style preference, but I'm a fan of `return`ing at the end of an `if` branch if that's all the processing that will happen in the method.  It reduces nesting of `if`/`else` clauses and sometimes helps readability.

File: lib/vmpooler/pool_manager.rb [hunk 20]
```diff
@@ -493,82 +577,96 @@ def _check_pool(pool)
       end
 
       # RUNNING
-      $redis.smembers('vmpooler__running__' + pool['name']).each do |vm|
+      $redis.smembers("vmpooler__running__#{pool['name']}").each do |vm|
         if inventory[vm]
           begin
-            check_running_vm(vm, pool['name'], $redis.hget('vmpooler__vm__' + vm, 'lifetime') || $config[:config]['vm_lifetime'] || 12)
+            check_running_vm(vm, pool['name'], $redis.hget('vmpooler__vm__' + vm, 'lifetime') || $config[:config]['vm_lifetime'] || 12, vsphere)
```
Comment (rick) [inline lib/vmpooler/pool_manager.rb]: Might pull the embedded `$redis.hget('vmpooler__vm__' + vm, 'lifetime') || $config[:config]['vm_lifetime'] || 12` out and assign it to a variable, for readability.

File: lib/vmpooler/pool_manager.rb [hunk 21]
```diff
@@ -493,82 +577,96 @@ def _check_pool(pool)
...
         end
       end
 
       # DISCOVERED
-      $redis.smembers('vmpooler__discovered__' + pool['name']).each do |vm|
-        %w(pending ready running completed).each do |queue|
-          if $redis.sismember('vmpooler__' + queue + '__' + pool['name'], vm)
-            $logger.log('d', "[!] [#{pool['name']}] '#{vm}' found in '#{queue}', removed from 'discovered' queue")
-            $redis.srem('vmpooler__discovered__' + pool['name'], vm)
+      discovered = $redis.smembers("vmpooler__discovered__#{pool['name']}")
```
Comment (rick) [inline lib/vmpooler/pool_manager.rb]: Could be inlined as `if discovered = ....` instead of split across two lines.  💄

File: spec/helpers.rb [hunk 22]
```diff
@@ -50,13 +50,25 @@ def create_pending_vm(template, name, token = nil)
   redis.hset("vmpooler__vm__#{name}", "template", template)
 end
 
-def create_vm(name, token = nil)
+def create_vm(name, token = nil, r = nil)
+  redis = r if r
```
Comment (rick) [inline spec/helpers.rb]: This is a bit tricky (I think we're shadowing a method with a local variable?).  Might be less surprising as:

``` ruby
def create_vm(name, token = nil, redis_handle = nil)
  redis_db = redis_handle ? redis_handle : redis
  redis_db.hset("vmpooler__vm__#{name}", 'checkout', Time.now)
  redis_db.hset("vmpooler__vm__#{name}", 'token:token', token) if token
end
```

File: spec/helpers.rb [hunk 23]
```diff
@@ -50,13 +50,25 @@ def create_pending_vm(template, name, token = nil)
...
   redis.hset("vmpooler__vm__#{name}", 'checkout', Time.now)
   if token
     redis.hset("vmpooler__vm__#{name}", 'token:token', token)
   end
 end
 
+def create_migrating_vm(name, pool, r = nil)
+  redis = r if r
```
Comment (rick) [inline spec/helpers.rb]: Ditto here, with redis overriding.

File: spec/helpers.rb [hunk 24]
```diff
@@ -50,13 +50,25 @@ def create_pending_vm(template, name, token = nil)
...
+  redis.hset("vmpooler__vm__#{name}", 'checkout', Time.now)
+  redis.sadd("vmpooler__migrating__#{pool}", name)
+end
+
+def add_vm_to_migration_set(name, r=nil)
+  redis = r if r
```
Comment (rick) [inline spec/helpers.rb]: Ditto here, with redis overriding.

File: spec/vmpooler/pool_manager_migration_spec.rb [hunk 25]
```diff
@@ -0,0 +1,93 @@
+require 'spec_helper'
+require 'mock_redis'
+require 'time'
+
+describe 'Pool Manager' do
+  let(:logger) { double('logger') }
+  let(:redis) { MockRedis.new }
+  #let(:redis) { Redis.new }
```
Comment (rick) [inline spec/vmpooler/pool_manager_migration_spec.rb]: Git has your back; feel free to drop commented bits.

File: spec/vmpooler/pool_manager_migration_spec.rb [hunk 26]
```diff
@@ -0,0 +1,93 @@
...
+  let(:pool) { config[:config][:pools][0]['name'] }
+  let(:vm) {
+    {
+      'name' => 'vm1',
+      'host' => 'host1',
+      'template' => pool,
+    }
+  }
+
+  before(:each) do
```
Comment (rick) [inline spec/vmpooler/pool_manager_migration_spec.rb]: This might not be necessary any more, since you're using `MockRedis`.

File: spec/vmpooler/pool_manager_migration_spec.rb [hunk 27]
```diff
@@ -0,0 +1,93 @@
...
+  end
+
+  describe '#_migrate_vm' do
+    let(:vsphere) { double(pool) }
+    let(:pooler) { Vmpooler::PoolManager.new(config, logger, redis, metrics) }
+    context 'evaluates VM for migration and logs host' do
+      before do
+        create_migrating_vm vm['name'], pool, redis
+        allow(vsphere).to receive(:find_vm).and_return(vm)
+        allow(pooler).to receive(:get_vm_host_info).and_return([{'name' => 'host1'}, 'host1'])
```
Comment (rick) [inline spec/vmpooler/pool_manager_migration_spec.rb]: Is there any way to let the redis state (presumably set up via `create_migrating_vm`) allow us to get away from mocking this call (and similar), and just exercise the db?  Or is there too much complexity behind this?

File: lib/vmpooler/pool_manager.rb (hunk 21 above)
Comment (mattkirby) [inline lib/vmpooler/pool_manager.rb]: Whoops. I'll fix it. This is some residual mess from me playing with re-arranging somethings. Thanks for catching it.

File: lib/vmpooler/pool_manager.rb (hunk 18 above)
Comment (mattkirby) [inline lib/vmpooler/pool_manager.rb]: `scard` definitely works in place of `smembers().size` here. It looks like that's what I should have used, and it has now been updated.

File: lib/vmpooler/pool_manager.rb (hunk 19 above)
Comment (mattkirby) [inline lib/vmpooler/pool_manager.rb]: That seems reasonable and like it helps make more clear the intent. I have added this where it looks appropriate to me in this method. It looks like I can just place `return` alone and have it return `nil`. Is that a reasonable convention, or is it preferred to return `true`?

File: spec/vmpooler/pool_manager_migration_spec.rb (hunk 27 above)
Comment (rick) [inline spec/vmpooler/pool_manager_migration_spec.rb]: I think it's fine. I'm willing to add a bit more technical debt here (with the notion that One Day™ we'll clean up the pool manager wrt better encapsulation and testability). Just wanted to make sure that we picked the low hanging :apple: if it were there, but sounds like (not surprisingly) it's not.

File: lib/vmpooler/pool_manager.rb (hunk 20 above)
Comment (mattkirby) [inline lib/vmpooler/pool_manager.rb]: I agree. Changing it.

File: spec/vmpooler/pool_manager_spec.rb (hunk 1 above)
Comment (rick) [inline spec/vmpooler/pool_manager_spec.rb]: I went through some months ago and converted most of our specs away from mocking/stubbing on redis, etc.  I stopped short on the pool_manager specs because there's just a lot of mess in that library and it needs deeper work (I think) to clean things up. 

I'm fine here with continuing in the existing style (because the cost of not doing so is large), even if the tests end up less than ideal: brittle, not fully exercising functionality.

**Comment by rick** [comment]
A few comments inline, but nothing really blocking. If this has been tested to behave as desired, I'm :+1:

File: spec/vmpooler/pool_manager_migration_spec.rb (hunk 27 above)
Comment (mattkirby) [inline spec/vmpooler/pool_manager_migration_spec.rb]: I do think some of the components I have re-arranged here will lend themselves to this being easier test. If it's valuable I'm happy to follow up on this to clean up pool manager further and implement tests around the rest of it as part of a future change.

Also, I think that these changes set us up to be able to actually test vsphere helper, which did not look very possible before.

**Comment by mattkirby** [comment]
It looks like my changes to pool manager have caused the original pool manager mockist tests to fail. I will need to update these too so checks pass.

**Comment by rick** [comment]
Tests passing!  🍏  💚  📗  :tea: 🇬🇱  ♻️

**Comment by rick** [comment]
I'm :+1:

**Comment by rick** [comment]
:+1: on the lost VM changes.

File: lib/vmpooler/pool_manager.rb [hunk 28]
```diff
@@ -40,25 +39,38 @@ def _check_pending_vm(vm, pool, timeout, vsphere)
...
-
-        $redis.smove('vmpooler__pending__' + pool, 'vmpooler__completed__' + pool, vm)
+    def remove_nonexistent_vm(vm, pool)
+      $redis.srem("vmpooler__pending__#{pool}", vm)
+      $logger.log('d', "[!] [#{pool}] '#{vm}' no longer exists. Removing from pending.")
+    end
 
-        $logger.log('d', "[!] [#{pool}] '#{vm}' marked as 'failed' after #{timeout} minutes")
+    def fail_pending_vm(vm, pool, timeout, exists=true)
+      begin
```
Comment (mattkirby) [inline lib/vmpooler/pool_manager.rb]: That's handy. I have updated the fail_pending_vm method accordingly. Let me know if it doesn't look right.

**Comment by rick** [comment]
:+1: on `rescue`-related changes.

File: lib/vmpooler/vsphere_helper.rb [hunk 29]
```diff
@@ -6,14 +6,23 @@ class VsphereHelper
...
 
-    def initialize(_vInfo = {})
-      config_file = File.expand_path('vmpooler.yaml')
-      vsphere = YAML.load_file(config_file)[:vsphere]
-
-      @connection = RbVmomi::VIM.connect host: vsphere['server'],
-                                         user: vsphere['username'],
-                                         password: vsphere['password'],
-                                         insecure: true
+    def initialize(credentials)
```
Comment (mattkirby) [inline lib/vmpooler/vsphere_helper.rb]: One benefit to this change is that connections are no longer made to vsphere when a thread is initialized. Instead, they are created when work is performed. The result is the application and its threads are available faster, and we do not open as many connections as pools when the application starts.

File: lib/vmpooler/pool_manager.rb [hunk 30]
```diff
@@ -20,39 +20,62 @@ def initialize(config, logger, redis, metrics)
...
-          Timeout.timeout(5) do
-            TCPSocket.new vm, 22
-          end
-          move_pending_vm_to_ready(vm, pool, host)
-        rescue
-          fail_pending_vm(vm, pool, timeout)
+          yield sock if block_given?
+        ensure
+          sock.close
+          return true
```
Comment (joshcooper) [inline lib/vmpooler/pool_manager.rb]: I think you'll have fewer changes if this method either yields the socket or raises an exception (so the caller doesn't need to check the return value)

File: lib/vmpooler/pool_manager.rb [hunk 31]
```diff
@@ -20,39 +20,62 @@ def initialize(config, logger, redis, metrics)
...
-          (((Time.now - Time.parse(clone_stamp)) / 60) > timeout)
+      if ! host
+        fail_pending_vm(vm, pool, timeout, false)
+        return
+      end
+      socket = open_socket vm
+      fail_pending_vm(vm, pool, timeout) if ! socket
+      move_pending_vm_to_ready(vm, pool, host)
+    rescue
+      fail_pending_vm(vm, pool, timeout)
```
Comment (joshcooper) [inline lib/vmpooler/pool_manager.rb]: for example this could just be:

```ruby
  open_socket vm
  move_pending_vm_to_ready(vm, pool, host)
rescue
  fail_pending_vm(vm, pool, timeout)
```

File: lib/vmpooler/pool_manager.rb (hunk 30 above)
Comment (mattkirby) [inline lib/vmpooler/pool_manager.rb]: You're right. I have updated it and it seems more clear this way.

**Comment by mattkirby** [comment]
I think this latest commit should address the suggestions in the last few comments @joshcooper.

**Comment by mattkirby** [comment]
I would like to propose merging this on 11/22 at 11am PST if there are no further issues to be raised here. Please let me know if this seems unreasonable and if additional changes are required before merging this.

I plan to rebase the changes before merging, but will hold off for tomorrow morning in order to provide reasonable time for feedback.

- [ ] Rebase commits to clean up commit history in preparation for merge.
- [ ] Wait until 11am PST to provide the opportunity for any further objections to be raised

**Comment by rick** [comment]
Just to double-check: have you been able to deploy this successfully to vmpooler-dev and observe the migration?

**Comment by mattkirby** [comment]
@rick, yes. The pre-tests incarnation of these changes are running on vmpooler-dev, vmpooler-cinext, and vmpooler-provisioner-dev-2 and have been running for the past several weeks. The latest commit here is running on vmpooler-provisioner-dev-2. I am happy to deploy the latest commit here to dev and vmpooler-cinext before deploying to production if that is helpful as well. My plan was otherwise to update each of those to be running against the merged commit prior to updating production vmpooler.

**Comment by rick** [comment]
Sounds great to me :+1:

**Comment by mattkirby** [comment]
I have rebased commits to clean up the history in preparation for merge.
